src/daft-recordbatch/src/ops/window.rs

```diff
@@ -89,8 +89,139 @@ impl RecordBatch {
+                WindowBoundary::UnboundedFollowing() => Some(total_rows),
+                WindowBoundary::UnboundedPreceding() => {
+                    return Err(DaftError::ValueError(
+                        "UNBOUNDED PRECEDING is not valid as an ending frame boundary".into(),
+                    ));
+                }
+                _ => None,
+            }
+        } else {
+            None
+        };
+
+        self.window_agg_non_incremental(
+            agg_expr,
+            &name,
+            dtype,
+            frame,
+            static_start,
+            static_end,
+            min_periods,
+            total_rows,
+        )
+    }
+
+    #[allow(clippy::too_many_arguments)]
+    fn window_agg_non_incremental(
+        &self,
+        agg_expr: &AggExpr,
+        name: &str,
+        dtype: &DataType,
+        frame: &WindowFrame,
+        static_start: Option<usize>,
+        static_end: Option<usize>,
+        min_periods: i64,
+        total_rows: usize,
+    ) -> DaftResult<Self> {
+        let null_series = Series::full_null(name, dtype, 1);
+
+        // Use the non-optimized implementation (recalculate for each row)
+        let mut result_series: Vec<Series> = Vec::with_capacity(total_rows);
+
+        for row_idx in 0..total_rows {
+            // Calculate frame bounds for this row
+            let frame_start = if let Some(idx) = static_start {
+                idx
+            } else {
+                match &frame.start {
+                    WindowBoundary::Offset(offset) => {
+                        (row_idx as i64 + offset).max(0).min(total_rows as i64) as usize
+                    }
+                    _ => unreachable!("Start boundary type already checked"),
+                }
+            };
+
+            let frame_end = if let Some(idx) = static_end {
+                idx
+            } else {
+                match &frame.end {
+                    WindowBoundary::Offset(offset) => ((row_idx + 1) as i64 + offset)
+                        .max(0)
+                        .min(total_rows as i64)
+                        as usize,
+                    _ => unreachable!("End boundary type already checked"),
+                }
+            };
+
+            let frame_size = frame_end as i64 - frame_start as i64;
+
+            if frame_size < 0 {
+                return Err(DaftError::ValueError(
+                    "Negative frame size is not allowed".into(),
+                ));
+            }
+
+            // Check min_periods requirement
+            if frame_size < min_periods {
+                // Add a null series for this row
+                result_series.push(null_series.clone());
+                continue;
+            }
+
+            // Calculate aggregation for this frame
+            let frame_data = self.slice(frame_start, frame_end)?;
+            let agg_result = frame_data.eval_agg_expression(agg_expr, None)?;
+            result_series.push(agg_result);
```

Comment:
Small nit for nicer flow of the code:

```suggestion
            if frame_size < min_periods {
                // Add a null series for this row
                result_series.push(null_series.clone());
            } else {
                // Calculate aggregation for this frame
                let frame_data = self.slice(frame_start, frame_end)?;
                let agg_result = frame_data.eval_agg_expression(agg_expr, None)?;
                result_series.push(agg_result);
            }
```